### ISSUE TITLE
host vars can't refer to other host vars of the same host when `hostvars` is used in a lookup

### ISSUE THREAD

**Issue by x-yuri** [ansible/ansible#43989]
##### SUMMARY

No way to use hostvars variable in subelements lookup when host vars refer to variables of the same host.

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

hostvars

##### ANSIBLE VERSION

```
ansible 2.6.2
  config file = /home/yuri/.ansible.cfg
  configured module search path = ['/home/yuri/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.6.6 (default, Jun 27 2018, 13:11:40) [GCC 8.1.1 20180531]
```

##### CONFIGURATION

RETRY_FILES_ENABLED(/home/yuri/.ansible.cfg) = False

##### OS / ENVIRONMENT

N/A

##### STEPS TO REPRODUCE

`playbook.yml`:

```yaml
- hosts: all
  tasks:
    - name: ...
      debug:
        msg: "{{ item.0.inventory_hostname }} {{ item.1 }}"
      with_subelements:
        - "{{ hostvars }}"
        - key1
```

`hosts`:

```yaml
h1
h2
```

`host_vars/h1/vars`:

```yaml
ansible_user: root
ansible_host: ...

key1: [1]
```

`host_vars/h2/vars`:

```yaml
ansible_user: root
ansible_host: ...

key1: [2]
vault_passwd: "{{ passwd }}"
passwd: 3
```

Archive with the [files][1].

[1]: https://github.com/ansible/ansible/files/2279715/1.tar.gz

##### EXPECTED RESULTS

`ansible-2.5.3`:

```
$ ansible-playbook -i hosts -l h1 playbook.yml

PLAY [all] ********************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************
ok: [h1]

TASK [...] ********************************************************************************************************************
ok: [h1] => (item=None) => {
    "msg": "h1 1"
}
ok: [h1] => (item=None) => {
    "msg": "h2 2"
}

PLAY RECAP ********************************************************************************************************************
h1                         : ok=2    changed=0    unreachable=0    failed=0
```

##### ACTUAL RESULTS

```
ansible-playbook -i hosts -l h1 playbook.yml

PLAY [all] ********************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************
ok: [h1]

TASK [...] ********************************************************************************************************************
fatal: [h1]: FAILED! => {"msg": "'passwd' is undefined"}

PLAY RECAP ********************************************************************************************************************
h1                         : ok=1    changed=0    unreachable=0    failed=1
```

##### MORE INFO

In `ansible-2.5` when `hostvars` was [rendered][mi1], it in its turn rendered variables of every host [right away][mi2]. And every other host variable was [available][mi3]. That is, given the following host vars:

[mi1]: https://github.com/ansible/ansible/blob/v2.5.7/lib/ansible/vars/hostvars.py#L118
[mi2]: https://github.com/ansible/ansible/blob/v2.5.7/lib/ansible/vars/hostvars.py#L92

[mi3]: https://github.com/ansible/ansible/blob/v2.5.7/lib/ansible/vars/hostvars.py#L91

```yaml
vault_passwd: "{{ passwd }}"
passwd: 3
```

When `"{{ passwd }}"` was renedered, `passwd` was among the available variables of the template.

In `ansible-2.6`, host variables are not [rendered][mi5] [right away][mi6]. `hostvars['h2']['vault_passwd']` would get you [`3`][mi7], but `repr(hostvars)` has [`"{{ passwd }}"`][mi8] in its place.

[mi5]: https://github.com/ansible/ansible/blob/v2.6.2/lib/ansible/vars/hostvars.py#L116
[mi6]: https://github.com/ansible/ansible/blob/v2.6.2/lib/ansible/vars/hostvars.py#L91
[mi7]: https://github.com/ansible/ansible/blob/v2.6.2/lib/ansible/vars/hostvars.py#L127-L128
[mi8]: https://github.com/ansible/ansible/blob/v2.6.2/lib/ansible/vars/hostvars.py#L142

**UPD**

The same goes for the `items` lookup,

```yaml
    - name: ...
      debug:
        msg: "{{ item.inventory_hostname }}"
      with_items:
        - "{{ hostvars }}"
```

But this way it iterates over `hostvars`'s keys,

```yaml
    - name: ...
      debug:
        msg: "{{ item }}"
      with_items: "{{ hostvars }}"
```

Supposedly since no flattening happens.

**The workaround**

```yaml
    - name: ...
      debug:
        msg: "{{ hostvars[item].inventory_hostname }} {{ hostvars[item].key1 }}"
      with_items: "{{ hostvars }}"
```

With `subelements` lookup the only solution is probably to move the task into a separate file.

**UPD**

There *is* a workaround. First I came up with a way to force it to render host vars,

```yaml
      with_subelements:
        - "{{ hostvars | force_render_hostvars }}"
        - key1
```

`filter_plugins/default.py`,

```py
class FilterModule(object):
    def filters(self):
        return {
            'force_render_hostvars': force_render_hostvars,
        }

def force_render_hostvars(hostvars):
    return {host: {k: v for k, v in hostvars[host].items()} for host in hostvars}
```

But it produces a lot of output. And then a better solution came to mind,

```yaml
      with_subelements:
        - "{{ hostvars | selectdictattr('key1') }}"
        - key1
```

`filter_plugins/default.py`,

```py
class FilterModule(object):
    def filters(self):
        return {
            'selectdictattr': selectdictattr,
        }

def selectdictattr(d, attr):
    return {k: {k2: v2 for k2, v2 in v.items() if k2 == attr} for k, v in d.items()}
```

The same goes for `items` lookup, just select the keys you really need.

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/vars/hostvars.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/vars/hostvars.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by fgtham** [comment]
This bites me as well. I use `hostvars` in `with_items` lookups alot. Upgrading to ansible 2.6 or 2.7 breaks all affected playbooks.

For now I am stuck with ansible 2.5.x.

**Comment by x-yuri** [comment]
But the story doesn't end there.

```yaml
a: '{{ b }}'
b: 1
```

```yaml
- name: ...
  debug:
    msg: '{{ hostvars["h1"]['a'] }}'
```

That gives 1, so far so good. Now then,

```yaml
a: '{{ hostvars["h1"].b }}'
b: 1
```

This time we get `'{{ hostvars["h1"].b }}'`.

In the first case it was able to get away with using the same [templar][1]. In the second one, it does [full][2] [rendering][3], but with variables of the host `h1` [only][4]. So, no `hostvars` variable available. It [fails][5], and as such returns the template itself. If you were to put this into a variable and use it somewhere (e.g. `set_fact` + `debug`), the rendering would succeed. But let's take it a bit more :D

```yaml
a: '{{ hostvars[c].b }}'
b: 1
c: 'h1'
```

[1]: https://github.com/ansible/ansible/blob/v2.7.4/lib/ansible/template/__init__.py#L458-L466
[2]: https://github.com/ansible/ansible/blob/v2.7.4/lib/ansible/template/__init__.py#L484-L491
[3]: https://github.com/ansible/ansible/blob/v2.7.4/lib/ansible/template/__init__.py#L735
[4]: https://github.com/ansible/ansible/blob/v2.7.4/lib/ansible/vars/hostvars.py#L127
[5]: https://github.com/ansible/ansible/blob/v2.7.4/lib/ansible/template/__init__.py#L770

**Comment by rawbertl** [comment]
Is there any solution for this using sole Jinja templates?

Given the following group-var:
`a: "{{ hostvars['somehost'].someVar }}"`

and the following Jinja template code:
...
`{{ hostvars['aHost']['a'] }}`

... it doesn't render the template and it won't output the actual content of 'someVar'. I haven't found any workaround yet which is quite a PITA... :-(

**Comment by x-yuri** [comment]
@rawbertp 

> If you were to put this into a variable and use it somewhere (e.g. set_fact + debug), the rendering would succeed.

```yaml
- name: ...
  set_fact:
    v1: "{{ hostvars['aHost']['a'] }}"
- name: ...
  debug:
    msg: "{{ v1 }}"
```

That is, you've got to render it twice. First it renders `{{ hostvars['aHost']['a'] }}` in the "global context." So `hostvars` is available. That gets us `{{ hostvars['somehost'].someVar }}`. Which is rendered in the context of the host (`hostvars` is not available). So it returns the template itself.

Then `debug` renders it again in the "global context", and it succeeds.

**Comment by rawbertl** [comment]
Thanks x-yuri. I noticed that it works with using set_fact but I'm afraid it won't work for my use case which looks as follows.

I'm using 'dynamically' generated variables which can be considered as the content of a properties file. This looks something like that:

```
prefix_property_a: valA
prefix_property_b: ValB
...
```
The result would be the following properties file:

```
property.a=valA
property.b=valB
...
```

To achieve this I iterate over all the hostvars (in my Jinja template):

```
% for var in hostvars[inventory_hostname]|sort %}{% if var.startswith( "prefix_") and hostvars[inventory_hostname][var] != None %}
{{ var['prefix_'|length:]|replace('___', '-')|replace('__', '<UNDERSCORE>')|replace('_', '.')|replace('<UNDERSCORE>', '_') }} = {{ hostvars[inventory_hostname][var] }}
{% endif %}{% endfor %}
```

While this works fine for the abovementioned example it doesn't work any longer if I set e.g. `prefix_property_b: hostvars['foo']['bar']`.

**Comment by muratdx** [comment]
> While this works fine for the abovementioned example it doesn't work any longer if I set e.g. prefix_property_b: hostvars['foo']['bar'].

Not sure if this is directly applicable to your use case, but I've been able to work around quirks in the new hostvars behavior using a hack like this: 

```hostvars[inventory_hostname] | to_json | from_json```

This appears to force rendering of all the template expressions for that host's hostvars collection.

**Comment by x-yuri** [comment]
@muratdx Your trick works in the following case:

`host_vars/h1`:

```yaml
a: 1
b: '{{ a }}'
```

`playbook.yml`:

```yaml
- hosts: all
  tasks:
    - name: ...
      debug:
        msg: '{{ hostvars[inventory_hostname] | to_json | from_json }}'
```

But not with:

```yaml
a: 1
b: '{{ hostvars['h1']['a'] }}'
```

Since `{{ hostvars['h1']['a'] }}` gets rendered with only `h1`'s vars, no `hostvars` variable availble. But if you were to put `{{ hostvars[inventory_hostname] }}` into, say, variable `v`, and do `{{ v }}`, it'd be rendered with "global" variables, including `hostvars`. So, it won't work in rawbertp's case.

But double rendering doesn't work with the first case's vars, when you get hostvars of the other host. Since the first rendering gets us `...b: '{{ a }}'...`, and then gets rendered with "global" (current host's) vars. That is, either `a` is taken from current host's vars, or unavailable.

@rawbertp You just have to render it twice.

`host_vars/h1`:

```yaml
prefix_property_a: valA
prefix_property_b: "{{ hostvars['h1']['prefix_property_a'] }}"
```

`playbook.yml`:

```yaml
- hosts: all
  tasks:
    - name: ...
      set_fact:
        v: |
            {% for var in hostvars[inventory_hostname]|sort %}{% if var.startswith( "prefix_") and hostvars[inventory_hostname][var] != None %}
                {{ var['prefix_'|length:]|replace('___', '-')|replace('__', '<UNDERSCORE>')|replace('_', '.')|replace('<UNDERSCORE>', '_') }} = {{ hostvars[inventory_hostname][var] }}
            {% endif %}{% endfor %}
    - name: ...
      debug:
        msg: '{{ v }}'
```

```
$ ansible-playbook -i h1, playbook.yml -vv
```

Or create a list of

```yaml
{{ var['prefix_'|length:]|replace('___', '-')|replace('__', '<UNDERSCORE>')|replace('_', '.')|replace('<UNDERSCORE>', '_') }} = {{ hostvars[inventory_hostname][var] }}
```

in one task, and `join` it in the other. Or create a dict with key:

```yaml
{{ var['prefix_'|length:]|replace('___', '-')|replace('__', '<UNDERSCORE>')|replace('_', '.')|replace('<UNDERSCORE>', '_') }}
```

and value:

```yaml
{{ hostvars[inventory_hostname][var] }}
```

And convert it into a string in the following task.

Replace `debug` task with task that puts it into a file.

**Comment by x-yuri** [comment]
@rawbertp You can even load template from a file, if that's what makes you think it won't work in your case:

`templates/1.j2`:

```yaml
{% for var in hostvars[inventory_hostname]|sort %}{% if var.startswith( "prefix_") and hostvars[inventory_hostname][var] != None %}
    {{ var['prefix_'|length:]|replace('___', '-')|replace('__', '<UNDERSCORE>')|replace('_', '.')|replace('<UNDERSCORE>', '_') }} = {{ hostvars[inventory_hostname][var] }}
{% endif %}{% endfor %}
```

`playbook.yml`:

```yaml
- hosts: all
  tasks:
    - name: ...
      set_fact:
        v: "{{ lookup('template', 'templates/1.j2') }}"
    - name: ...
      debug:
        msg: "{{ v }}"
```

**Comment by rawbertl** [comment]
Hi @x-yuri  - thank you for the valuable hints, much appreciated! I will have to see what works best for me. I'm afraid will have to find a solution without setting a fact though,  due to situations were mutltiple templates will be generated in parallel (on the same host but for different applications) resulting in one execution overwriting the fact of another...

(In the actual implementation it is not `var['prefix_']...` but `var[app_name + '_prefix_']...` allowing application-specific properties.)

**Comment by x-yuri** [comment]
@rawbertp This is probably starting to get off topic, but... You can always create a dict (`{app1: {var1: val1, var2: val2, ...}, app2: {...}, ...}`). The critical part (that can't be done in external template) is `hostvars[inventory_hostname][var]`. So, I'd first create such dict with all the values evaluated (templates rendered), then pass it to `template` module. Also, why don't you save yourself some effort and create the dict right in the host vars?

```yaml
appvars:
  app1:
    var1: val1
...
```

**Comment by rawbertl** [comment]
>  Also, why don't you save yourself some effort and create the dict right in the host vars?

I know that would be the easiest solution but unfortunately doesn't work for my use case. The app properties can originate from multiple sources/repositories. So the final Ansible project is composed of various sources which would not work with a dict. (at least not to my knowledge)

Sorry for the off-topic! 


---

Update: I just figured out that the following workaround also appears to do the trick.

1. Execute the template module as local action
2. Execute the template module again with src == result of (1)

**Comment by samdoran** [comment]
@sivel Is this related to deepdish for 2.6? Is this something we want to fix?

**Comment by s-hertel** [comment]
I was able to reproduce this on 2.6 and it's been fixed in recent versions with #64282. I wasn't able to replicate all comments, so if you're experiencing a similar issue on recent versions, please open a new issue with a reproducer.
